cmd/machine-os-builder/start.go

```diff
@@ -89,25 +95,50 @@ func runStartCmd(_ *cobra.Command, _ []string) {
 	// To help debugging, immediately log version
 	klog.Infof("Version: %+v (%s)", version.Raw, version.Hash)
 
+	// This is the 'main' context that we thread through the build controller context and
+	// the leader elections. Cancelling this is "stop everything, we are shutting down".
 	ctx, cancel := context.WithCancel(context.Background())
-
 	cb, err := clients.NewBuilder("")
 	if err != nil {
 		klog.Fatalln(err)
 	}
 
-	ctrl, err := getBuildController(ctx, cb)
-	if err != nil {
-		klog.Fatalln(err)
-		var invalidImageBuiler *build.ErrInvalidImageBuilder
-		if errors.As(err, &invalidImageBuiler) {
-			klog.Errorf("The user passed an invalid imageBuilderType of %s", invalidImageBuiler.InvalidType)
-			cancel()
-			os.Exit(255)
+	run := func(ctx context.Context) {
+		go common.SignalHandler(cancel)
+
+		controllers, err := getBuildControllers(ctx, cb)
+		if err != nil {
+			klog.Fatalln(err)
+			var invalidImageBuiler *build.ErrInvalidImageBuilder
+			if errors.As(err, &invalidImageBuiler) {
+				klog.Errorf("The user passed an invalid imageBuilderType of %s", invalidImageBuiler.InvalidType)
+				cancel()
+				os.Exit(255)
+			}
+		}
+		for _, ctrl := range controllers {
+			go ctrl.Run(ctx, 3)
 		}
+		<-ctx.Done()
+		cancel()
 	}
 
-	go ctrl.Run(ctx, 5)
-	<-ctx.Done()
-	cancel()
+	leaderElectionCfg := common.GetLeaderElectionConfig(cb.GetBuilderConfig())
+
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:            common.CreateResourceLock(cb, ctrlcommon.MCONamespace, componentName),
+		ReleaseOnCancel: true,
+		LeaseDuration:   leaderElectionCfg.LeaseDuration.Duration,
+		RenewDeadline:   leaderElectionCfg.RenewDeadline.Duration,
+		RetryPeriod:     leaderElectionCfg.RetryPeriod.Duration,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				klog.Infof("Stopped leading. MOB terminating.")
+				os.Exit(0)
+			},
+		},
+	})
+	panic("unreachable")
```

Comment:
suggestion (non-blocking): The `panic()` call is unnecessary and can be removed.